cranelift/reader/src/run_command.rs

```diff
@@ -0,0 +1,148 @@
+                write!(f, "run: {} {} ", invocation, comparison)?;
+                if expected.len() == 1 {
+                    write!(f, "{}", expected[0])
+                } else {
+                    write!(f, "[")?;
+                    write_data_value_list(f, expected)?;
+                    write!(f, "]")
+                }
+            }
+        }
+    }
+}
+
+/// Represent a function call; [RunCommand]s invoke a CLIF function using an [Invocation].
+#[derive(Debug, PartialEq)]
+pub struct Invocation {
+    /// The name of the function to call. Note: this field is for mostly included for informational
+    /// purposes and may not always be necessary for identifying which function to call.
+    pub func: String,
+    /// The arguments to be passed to the function when invoked.
+    pub args: Vec<DataValue>,
+}
+
+impl Invocation {
+    pub(crate) fn new(func: &str, args: Vec<DataValue>) -> Self {
+        let func = func.to_string();
+        Self { func, args }
+    }
+}
+
+impl Display for Invocation {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        write!(f, "%{}(", self.func)?;
+        write_data_value_list(f, &self.args)?;
+        write!(f, ")")
+    }
+}
+
+/// Represent a data value. Where [Value] is an SSA reference, [DataValue] is the type + value
+/// that would be referred to by a [Value].
+#[allow(missing_docs)]
+#[derive(Clone, Debug, PartialEq)]
+pub enum DataValue {
+    B(bool),
+    I8(i8),
+    I16(i16),
+    I32(i32),
+    I64(i64),
+    F32(f32),
+    F64(f64),
+    V128([u8; 16]),
+}
+
+/// Helper for creating [From] implementations for [DataValue]
+macro_rules! from_data {
+    ( $ty:ty, $variant:ident ) => {
+        impl From<$ty> for DataValue {
+            fn from(data: $ty) -> Self {
+                DataValue::$variant(data)
+            }
+        }
+    };
+}
+from_data!(bool, B);
+from_data!(i8, I8);
+from_data!(i16, I16);
+from_data!(i32, I32);
+from_data!(i64, I64);
+from_data!(f32, F32);
+from_data!(f64, F64);
+from_data!([u8; 16], V128);
+
+impl Display for DataValue {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        match self {
+            DataValue::B(dv) => write!(f, "{}", dv),
+            DataValue::I8(dv) => write!(f, "{}", dv),
+            DataValue::I16(dv) => write!(f, "{}", dv),
+            DataValue::I32(dv) => write!(f, "{}", dv),
+            DataValue::I64(dv) => write!(f, "{}", dv),
+            // Use the Ieee* wrappers here to maintain a consistent syntax.
+            DataValue::F32(dv) => write!(f, "{}", Ieee32::from(*dv)),
+            DataValue::F64(dv) => write!(f, "{}", Ieee64::from(*dv)),
+            // Again, for syntax consistency, use ConstantData, which in this case displays as hex.
+            DataValue::V128(dv) => write!(f, "{}", ConstantData::from(&dv[..])),
+        }
+    }
+}
+
+/// Helper function for displaying `Vec<DataValue>`.
+fn write_data_value_list(f: &mut Formatter<'_>, list: &[DataValue]) -> Result {
```

Comment:
Isn't this possibly inferred by the compiler for Vec<T> when T implements Display? Or can we use `{:?}` for the Vec<T> type and it'll do the right thing here?

Reply:
@alexcrichton suggested I use `impl Display` instead of `DataValue`, which is a good solution to avoid this function; but I remembered that I made this because sometimes I need the data values enclosed in parentheses and sometimes in brackets. E.g. `%fn() = [1, 2, 3]` vs `%fn(1, 2, 3) == 42`. I'm going to leave this function in for that reason.